server/monitor/credentials.go

```diff
@@ -0,0 +1,21 @@
+package monitor
+
+import (
+	"time"
+)
+
+func (m *Monitor) checkCredentials(force bool) {
+	if !force && (time.Now().Hour() != 0 || time.Now().Minute() != 0) {
+		return
+	}
+
+	credentials, err := m.client.GetAppCredentials(m.applicationID)
+	if err != nil {
+		m.api.LogDebug("Error getting client secret expire date", "error", err.Error())
+		return
+	}
+
+	for _, credential := range credentials {
+		m.metrics.ObserveClientSecretExpireDate(credential.ID, credential.ExpireDate)
+	}
```

Comment:
I realize this is likely to be finite and very small in practice, but now that I see the idea we discussed in practice, it feels potentially problematic to create a discrete time series for each `credential.ID`. 

Would we consider sorting the `credentials` by `ExpireDate` (ascending, so we see the expired / next to expire first), and then reporting on at most `5`? We could then name the labels programmatically as `credential_0`, bounding the total number of time series. As credentials are removed, we would see the times shift, but this is probably sufficient fidelity for our purposes.

Reply:
That sounds like some extra complexity that I don't see the need for. From my perspective you should only have 1 credential, and in the worst case scenario, you are going to renew that credentials every 3 months. So I don't think is going to be that many metrics. Can you elaborate more what is the problem of having multiple metrics? from my perspective, your are not even querying for the whole history, you are probably query for the current credentials, we can even stop reporting credentials that are expired for more than a month, or something like that (I don't think is needed anyway).

Reply:
Every new `credential.ID` creates a new Prometheus timeseries. As I understand it, this is like a brand new `CREATE TABLE`. Even if you don't add new data after a while, I understand it to be best practice to avoid a potentially unbounded input set.

You are right, @jespino, that in practice, this is unlikely to be problematic. I also wonder if maybe our idea of a metric for this was altogether unnecessary in the first place, given we have to ask customers to give the application extra permissions.

Thoughts about just abandoning this in favour of a plan to contact specific customers re: credential expirations, and do more battle testing on the plugin when those credentials stop working (plus resiliency to recover) so we can help customers who forget to transition across this threshold?

Reply:
I'm ok with both approaches, but I don't think that approach is very "forgiving", and is not very scalable. Having this properly monitored gives you a clear way to act on that. Not having that monitored makes this error-prone. And can end up leading to down times in the plugin.

I see the drawback on adding an extra permission requirement, and I see the drawback on adding one credential ID for certificate. But, I don't see the point on removing that. If you don't give the permission I'm not going to create any metric, so, that has no impact on prometheus, if you have the permission, the metric is going to be there and isgoing to be providing value.

Another option is to only show the one that is more "expired" or more near to expire. That way we can directly communicate that and they have to remove that to ensure that we are showing the right expire date.